Support Scribble Handwriting

fbcouch opened the pull request:
Engine PR: https://github.com/flutter/engine/pull/24224

This PR is the framework side of adding scribble handwriting support. I still need to add some tests, but looking for feedback on the overall approach. I incorporated some feedback from @LongCatIsLooong on a previous draft, moving most of the logic of this into the framework and adding a widget (`_ScribbleElement`) to handle most of the scribble-specific logic.

The one remaining issue here (that I could use some ideas on how to solve) is that the UIIndirectScribbleInteraction can always be triggered, regardless of whether the text input in question is obscured by another element or otherwise non-interactive (e.g. behind something else in a stack).

https://github.com/flutter/flutter/issues/61278

Design doc: https://docs.google.com/document/d/1mjQbsSRQnHuAgMNdouaSgTS-Xv-w57fdKfOUUafWpRo

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

LongCatIsLooong reviewed (commented):
Thank you for the PR and great work! Left some initial feedback and questions.

An additional question regarding the `UIIndirectScribbleInteractionDelegate` protocol:
when does UIKit call `func indirectScribbleInteraction(UIInteraction, frameForElement: Self.ElementIdentifier) -> CGRect`? Is it always immediately after one of the async calls (`func indirectScribbleInteraction(UIInteraction, requestElementsIn: CGRect, completion: ([Self.ElementIdentifier]) -> Void)` or `func indirectScribbleInteraction(UIInteraction, focusElementIfNeeded: Self.ElementIdentifier, referencePoint: CGPoint, completion: ((UIResponder & UITextInput)?) -> Void)`)? I'd like to understand how much information we would have to send to the engine every frame.

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -856,6 +867,27 @@ abstract class TextInputClient {
+  void onScribbleFocus(double x, double y);
+
+  /// Tests whether the [ScribbleClient] overlaps the given rectangle bounds
+  bool inScribbleRect(double x, double y, double width, double height);
```
I agree! Do you have any thoughts on the best way to figure that out? Between that and other things that might obscure the fields from being interactive, I think that is the biggest remaining issue.

I think the rest of your feedback all makes sense at first glance, thank you for taking the time to review this!

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2423,6 +2423,15 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       final Size size = renderEditable.size;
       final Matrix4 transform = renderEditable.getTransformTo(null);
       _textInputConnection!.setEditableSizeAndTransform(size, transform);
+      final TextSpan textSpan = buildTextSpan();
```
This is to get the length of the text for use on line 2430

LongCatIsLooong replied:
Oh I see to implement `closestPositionToPoint` and friends you will need the bounding boxes for each glyph. This is going to be pretty heavy for long documents, and since a `RenderEditable` clips its content if it gets too long (i.e. `RenderEditable` is a scrollable container and only part of it will remain visible), you'll get false positives on glyphs that are not currently visible.

LongCatIsLooong replied:
What scribble function(s) would it break if we leave the `closestPositionToPoint` family unimplemented?

fbcouch replied:
Oh interesting. I will have to put some long text in and see how that affects things.

So, `closestPositionToPoint` itself allows UIKit to move the selection around based on where the user is writing. That one is not a huge deal to leave unimplemented, I suppose – the caret will just stay where it is using the previous implementation. The bigger ones are `firstRectForRange` and `selectionRectsForRange` which UIKit needs to do pretty much everything except inserting text at the caret: selecting, deleting, inserting/removing spaces

fbcouch replied:
Okay, so you are right, text long enough to scroll does cause some issues. I pushed up some changes to deal with this:

1. Offset the selection rects by the current scroll offset
2. Cache the selection rects so that they are only recalculated if needed
3. Wait for the scroll direction to be `.idle` so that scrolling is smooth

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1236,10 +1281,24 @@ class TextInput {
   TextInputConnection? _currentConnection;
   late TextInputConfiguration _currentConfiguration;
 
+  final Map<String, ScribbleClient> _scribbleClients = <String, ScribbleClient>{};
```
Would a static member of `EditableTextState` make sense? Or maybe `_ScribbleElement` should become public and this and the register/unregister static methods would all live there?

fbcouch commented:
> Thank you for the PR and great work! Left some initial feedback and questions.
> 
> An additional question regarding the `UIIndirectScribbleInteractionDelegate` protocol:
> when does UIKit call `func indirectScribbleInteraction(UIInteraction, frameForElement: Self.ElementIdentifier) -> CGRect`? Is it always immediately after one of the async calls (`func indirectScribbleInteraction(UIInteraction, requestElementsIn: CGRect, completion: ([Self.ElementIdentifier]) -> Void)` or `func indirectScribbleInteraction(UIInteraction, focusElementIfNeeded: Self.ElementIdentifier, referencePoint: CGPoint, completion: ((UIResponder & UITextInput)?) -> Void)`)? I'd like to understand how much information we would have to send to the engine every frame.

I believe UIKit calls `func indirectScribbleInteraction(UIInteraction, requestElementsIn: CGRect, completion: ([Self.ElementIdentifier]) -> Void)` and waits for the completion handler. Then it calls `func indirectScribbleInteraction(UIInteraction, frameForElement: Self.ElementIdentifier) -> CGRect`. Since I have `func indirectScribbleInteraction(UIInteraction, shouldDelayFocusForElement: Self.ElementIdentifier) -> Bool` always returning true, it waits for a pause in the user input before calling `func indirectScribbleInteraction(_ interaction: UIInteraction, focusElementIfNeeded elementIdentifier: Self.ElementIdentifier, referencePoint focusReferencePoint: CGPoint, completion: @escaping ((UIResponder & UITextInput)?) -> Void)`

I don't think we're sending info to the engine on every frame during the indirect interaction now that I moved that to the framework side. Once the input is focused, we do calculate the selection boxes every frame, but we also cache those on the framework side so that they only actually get sent to the engine when changed.

LongCatIsLooong commented:
(sorry for the delay I'll take a look tonight).

fbcouch commented:
No problem! I actually have an idea for how to do the hit testing thing, just haven't had a chance to cherry pick that into the PR branch. That's here: https://github.com/twinsunllc/flutter/commit/0dfa712f079c1f308c75ab6a7c840ca6095750ac#diff-f5f93c879cef9a102adbf148583de9f3d9a05f9678fdcad7bde106f62200bb34R2947

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3064,66 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
+  RenderEditable? get renderEditable => widget.editableKey.currentContext?.findRenderObject() as RenderEditable?;
+
+  static int _nextElementIdentifier = 1;
+  String _elementIdentifier;
```
this can be final right?

LongCatIsLooong reviewed (commented):
Sorry I just realized I can't test this on a simulator. I'll see if I can get a device.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2418,11 +2418,40 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+      final Offset rectOffset = Offset(_isMultiline ? 0.0 : -1 * scrollOffset, _isMultiline ? -1 * scrollOffset : 0.0);
+      final List<Rect> rects = List<Rect>.generate(
+              text.length, (int i) => renderEditable.getBoxesForSelection(TextSelection(baseOffset: i, extentOffset: i + 1)).first.toRect())
+          .map((Rect rect) => rect.translate(rectOffset.dx, rectOffset.dy))
```
this should probably happen in the `RenderEditable` public api you exposed in the PR.

fbcouch replied:
So using `localToGlobal` on the render object there instead of offsetting by the scroll offset here?

LongCatIsLooong replied:
The `RenderEditable.getBoxesForSelection` API you exposed in the PR should be responsible for doing the `translate(rectOffset.dx, rectOffset.dy)` thing. The caller would usually expect the `TextBox`es to be described in the `RenderEditable`'s coordiante space instead of `_textPainter`'s.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2418,11 +2418,40 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+      _cachedFirstRect = firstRect;
+      final Offset rectOffset = Offset(_isMultiline ? 0.0 : -1 * scrollOffset, _isMultiline ? -1 * scrollOffset : 0.0);
+      final List<Rect> rects = List<Rect>.generate(
+              text.length, (int i) => renderEditable.getBoxesForSelection(TextSelection(baseOffset: i, extentOffset: i + 1)).first.toRect())
```
I'm not sure if this is going to work when there's surrogate pairs & extended glypheme clusters. @justinmc

LongCatIsLooong replied:
Will UIkit ever ask for the locations of the glyphs in a text field that is not focused? If that's the case the Rects can be outdated since we only update for the current text field.

fbcouch replied:
I believe it will not. When not focused, we should only be dealing with `UIIndirectScribbleInteraction` which asks for the elements in a rect (`requestElementsInRect`) and then lets us know to focus on one of them at a given point (`focusElementIfNeeded`). Once an element has focus, we should be sending over the rects so that UIKit has what it needs to do the text operations that happen during a `UIDirectScribbleInteraction`

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1236,10 +1276,47 @@ class TextInput {
+      }
+      return;
+    } else if (method == 'TextInputClient.requestElementsInRect') {
+      final List<dynamic> args = methodCall.arguments as List<dynamic>;
```
The idea being to move the `_scribbleClients` and `registerScribbleElement` / `unregisterScribbleElement` members into a `ScribbleContext` or some such thing?

fbcouch commented:
@LongCatIsLooong I think I have addressed your feedback except for a couple of items:

- https://github.com/flutter/flutter/pull/75472#discussion_r584234318 - not sure what to do here
- https://github.com/flutter/flutter/pull/75472#discussion_r584235087 - is there a way to test this? I believe Scribble only supports English and Chinese, so I'm not sure if we would run into those things

I also added an implementation for `insertTextPlaceholder` and `removeTextPlaceholder` coming from the engine side. PencilKit uses those to accomplish the Insert scribble method where you can long press with the pencil to get some extra space to write in.

Let me know your thoughts there. I have some merge conflicts now – I'm hoping this is close, but I'll wait to get your thumbs up before rebasing this onto the current master (or perhaps the last green commit on master?) and marking it as ready for review.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3119,135 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
+      intersection.bottomCenter,
+      intersection.bottomRight
+    ].any((Offset point) {
+      final HitTestResult result = HitTestResult();
```
Ah I see how difficult it would be to find hitable `RenderObject`s in a given `Rect`. Hitable `RenderBox`es can partially overlap each other. So never mind the previous comment about adding things to `RenderObject`. But still we probably shouldn't sample points. 

@goderbauer @justinmc any suggestions on how we can identify currently "interactable" render objects in a given rect?

goderbauer replied:
Yeah, sampling "random" points doesn't seem right here.

Also it seems we are doing this hit testing multiple times if there are multiple Scribbles on screen?

Can you describe what the hit test is trying to accomplish?

LongCatIsLooong replied:
@goderbauer there's an interesting scribble feature where you can scribble on screen without any text field focused. The text will be sent to the most relevant text field based on the scribble location. An example: in the "reminder" app you can launch the app and just scribble away, and a new "reminder" entry will be added.

To that end UIKit needs to ask us for "scribble elements" in a certain `Rect`. But it's tricky to tell if a `ScribbleClient` has a "hotspot" in the given a `Rect` (as it can be behind an `IgnorePointer` or obscured by another `ScribbleClient`, we would have to hit test every `Offset` in the `Rect`).

fbcouch replied:
Right, so the hit testing is last here, after making sure the `_ScribbleFocusable` intersects the rect that UIKit is interested in. I think in most cases that would just be a few elements. It also uses `any` so that in the nominal case where the whole text field is visible and interactive, only one point will need to be checked.

I am interested to hear some ideas on the best way to do this. FWIW I think this approach will work in most cases, since it is testing points on the intersection between the two rectangles. Unless someone has very large handwriting, that area should be relatively small.

We actually have a beta test going with a custom engine implementing this that seems to be going pretty well. Granted, that is on a separate branch as we haven't updated from flutter 1.22.2 – I will probably look into upgrading again once this makes it to the stable branch.

justinmc replied:
What if we just hit test the center of the intersection and that's it?  Are there any real cases where that would fail and the user would reasonably expect it to succeed?

Unless I'm missing some case, it seems like hit testing the center of the intersection would only fail if the EditableText is more than half way obscured.  In reality, if the EditableText is obscured, the user probably would forgive us for not entering their text.  For example, if the field is partially (at least halfway) scrolled underneath the AppBar, it would be clear that the user needs to reveal it first to get it to work.

The main cases where this approach wouldn't work would be crazy polygon shaped widgets on top of text fields, which probably isn't vital to get correct in a feature like this.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3119,135 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
+    final RenderBox? box = context.findRenderObject() as RenderBox?;
+    if (box == null || !mounted || !box.attached)
+      return Rect.zero;
+    final Offset topLeft = box.localToGlobal(Offset.zero);
```
This may give you the wrong size because only the origin is transformed. The transform can be retrieved using `getTransformTo` and it should be applied to the rect. There doesn't seem to a shorthand method for this unfortunately.

LongCatIsLooong reviewed (commented):
Great work! 

The "finding the right scribble clients" problem seems to be way harder than I initially thought and unfortunately I don't really have a good solution.

/cc  @justinmc for insights.

goderbauer commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3119,135 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
+  @override
+  void onScribbleFocus(Offset offset) {
+    widget.focusNode.requestFocus();
+    renderEditable?.selectPositionAt(from: offset, cause: SelectionChangedCause.keyboard);
```
cause: keyboard is odd. Is that correct here?

goderbauer commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3119,135 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
+  /// The size of the span, used in place of adding a placeholder size to the [TextPainter]
+  final Size size;
+
+  /// Adds a placeholder box to the paragraph builder if a size has been
```
no need to repeat the docs of an inherited method.

goderbauer reviewed (commented):
(not a full review)

fbcouch commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -62,6 +62,9 @@ enum SelectionChangedCause {
   /// of text.
   drag,
+
+  /// The user iPadOS 14 Scribble to change the selection.
```
Perhaps, but I don't know the behavior for those – for scribble specifically, this is needed so that we show the selection handles like native the iPadOS UITextField does when using scribble to change the selection...I don't know if that is the case for using a stylus to change the selection on android, for example

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -856,6 +856,33 @@ abstract class TextInputClient {
+  void showToolbar();
+
+  /// UIKit has requested a text placeholder be added at the current selection
+  void insertTextPlaceholder(Size size);
```
It does not appear that you can add multiple placeholders, that said, we do need to clean up the one in the current `TextInputClient` if the focus changes, since the `removeTextPlaceholder` call will get pushed to the new `TextInputClient`

fbcouch replied:
Oh, I forgot to answer your first question: this is another scribble feature where you can long press to give yourself extra space to write in the middle of some text. I don't know if it's used for anything else...the docs about it are pretty terse, and don't event mention the scribble use.

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -856,6 +856,33 @@ abstract class TextInputClient {
+  void removeTextPlaceholder();
+}
+
+/// An interface for recieving focus during a UIIndirectScribbleInteraction
```
Do you have any thoughts on what parts of this would be abstracted out? This is pretty closely tied to the way iPadOS's scribble feature interacts with the UITextInput, so I'm not sure how generic it would be

fbcouch replied:
I think I see what you mean – just keeping the docs at a more abstract level...I think I got all of those and made the docs more generic (e.g. handling x from engine) with the stuff about specific iOS features as a "for example"

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1236,10 +1282,47 @@ class TextInput {
+    if (method == 'TextInputClient.focusElement') {
+      final List<dynamic> args = methodCall.arguments as List<dynamic>;
+      if (_scribbleClients.containsKey(args[0])) {
+        _scribbleClients[args[0]]?.onScribbleFocus(Offset(args[1].toDouble() as double, args[2].toDouble() as double));
```
I think the issue is that `args` is `List<dynamic>`, so `args[1]` is `dynamic` which means `args[1].toDouble()` also returns dynamic. If i take out `as double` I get a the following error: 

> The argument type 'dynamic' can't be assigned to the parameter type 'double'.

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1236,10 +1282,47 @@ class TextInput {
+      }
+      return;
+    } else if (method == 'TextInputClient.requestElementsInRect') {
+      final List<dynamic> args = methodCall.arguments as List<dynamic>;
```
Maybe I'm doing something wrong, but I tried this `final List<double> args = methodCall.arguments as List<double>;` and ended up with:

> _CastError (type 'List<dynamic>' is not a subtype of type 'List<double>' in type cast)

justinmc replied:
I think you need [List.cast](https://master-api.flutter.dev/flutter/dart-core/List/cast.html).

fbcouch replied:
I see. I think the issue I was running into is that `arguments` is `dynamic`, so it needs to be cast to a list with `as List<dynamic>` _then_ I can call `.cast<double>()` to get what we want here

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2418,11 +2420,34 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+    textSpan.computeToPlainText(text);
+    final Rect firstRect = renderEditable.getBoxesForSelection(const TextSelection(baseOffset: 0, extentOffset: 1)).first;
+    final ScrollDirection scrollDirection = _scrollController?.position.userScrollDirection ?? ScrollDirection.idle;
+    if (scrollDirection == ScrollDirection.idle && (text.toString() != _cachedText ||
```
`TextSpan` has `computeToPlainText`, which accepts a `StringBuffer`, so `.toString()` here is actually called on `StringBuffer`

For `ScrollDirection`, the idea is to wait to send updated selection rects until scrolling stops – that way large amounts of text don't cause scrolling to be choppy.

That's a good point on the cache invalidation...the size is easy enough to check, I suppose, are there any other invalidation conditions you can think of?

LongCatIsLooong replied:
`toString` produces debug messages for the object so in the future more stuff might be added to it. I think `computeToPlainText` might be the more robust choice. 

> are there any other invalidation conditions you can think of

If the paragraph starts with a zero-width glyph (for example, line feed), it's not going to represent the layout of other glyphs well.

fbcouch replied:
Gotcha, well, it is using `computeToPlainText` so I think we're OK on that. For the zero width glyph...that might be OK, I think I was just using the firstRect as a stand in for the offset of the text, since if it gets scrolled, the rects will need to be recalculated. Maybe it's OK without that? I'll have to do a check with some longer text and see.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2519,6 +2545,24 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     }
   }
 
+  int _placeholderLocation = -1;
```
I actually kinda like using -1 as a special value here, as I can easily test that the placeholder is within the bounds of the string (which I need to do anyway) without having to also make sure `_placeholderLocation` is not null

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2519,6 +2545,24 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+  void insertTextPlaceholder(Size size) {
+    print('[scribble][flutter] insertTextPlaceholder $size');
+    setState(() {
+      _placeholderLocation = _value.text.length - widget.controller.selection.end;
```
Can the `EditableText` have focus but not a selection? The checks on where the placeholder gets inserted _should_ handle that (they test that it falls within [0, _value.text.length]), so we wouldn't insert the placeholder anywhere...which...if there's no selection, we wouldn't know where to put it anyway.

LongCatIsLooong replied:
Yeah the user can set the selection to `(-1, -1)` even if the `EditableText` is focused.

LongCatIsLooong replied:
You might want to check if the selection is valid (https://master-api.flutter.dev/flutter/dart-ui/TextRange/isValid.html) first.

LongCatIsLooong replied:
I'm curious how text selections interact with placeholders. When a placeholder is inserted and the current selection isn't empty, does the placeholder replace the currently selected text, or the placeholder is simply inserted at the end of the current selection? If it's the latter what happens if you press backspace after inserting a placeholder?

fbcouch replied:
The placeholder should be inserted wherever the user is pressing – the selection actually gets moved there and collapsed anyway when the pencil is placed.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3119,135 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
   }
 }
+
+class _ScribbleElement extends StatefulWidget {
```
Would `_ScribbleFocusable` be a better name? It is certainly more descriptive.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3053,3 +3119,135 @@ class _WhitespaceDirectionalityFormatter extends TextInputFormatter {
+
+  @override
+  bool isInScribbleRect(Rect rect) {
+    final Rect _bounds = bounds;
```
`bounds` is computed, so I figured it was worth getting once and setting it to a local variable to use in the later steps

justinmc reviewed (commented):
Quick review.  This is exciting!  Mainly see the comments about extended grapheme clusters.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2418,13 +2420,42 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+      _cachedFirstRect = firstRect;
+      _cachedSize = size;
+      final List<Rect> rects = List<Rect>.generate(
+              text.length, (int i) => renderEditable.getBoxesForSelection(TextSelection(baseOffset: i, extentOffset: i + 1)).first);
```
@justinmc I did some refactoring to use characters, let me know how that looks. It seems to work well with at least those specific multi-byte characters that you mentioned now.

fbcouch replied:
All right, so I did find one more issue that's pretty strange with characters like 👨‍👩‍👦 that contain zero width joiners. With the following text:

> 
> 👨‍👩‍👦 1 2 3 4 5 6 7 8

(Note the leading newline) – only "👨‍👩‍👦 1 2 3 4 5 " will be selectable with scribble. However, adding a trailing newline makes the whole line selectable.

I've been poking around various things and trying to compare them to the native UITextView versions of these, and really have no idea what's going on. It really looks like PencilKit basically searches (correctly) positions 1-24 at first, then comes back through and only looks at 1-20 before setting the selected text to 1-19. If you have any thoughts on why that might be, I'm all ears.

All that said, it works perfectly with 😆 in all of my testing, so I think this is limited to ZWJ-containing characters.

fbcouch replied:
@justinmc All right, I think the ZWJ-containing issue is resolved with the latest updates to the engine part of this PR

fbcouch commented:
@LongCatIsLooong @justinmc I believe I've addressed all of the comments here, let me know if you see anything else! Thanks!

fbcouch commented:
@LongCatIsLooong @justinmc @goderbauer Just checking in here if there are any more changes needed. It looks like there are some merge conflicts now, so if everything looks good, I can clean those up and hopefully we can get this merged 😄

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -908,6 +911,74 @@ abstract class TextInputClient {
+  /// [bounds].
+  const SelectionRect({required this.position, required this.bounds});
+
+  /// The position of this selection rect within the text.
```
Can you elaborate exactly what this position is in these docs?  Is it the baseOffset of the selection, or is it whichever comes first depending on the TextDirection?

fbcouch replied:
Should this go in the comment? I don't know how it would work with a rtl language (are Traditional Chinese or Cantonese right to left on iOS?), but this is essentially the base offset of a selection that would contain this SelectionRect, so for example, this will be `0` for the first character, `1` for the second, and so on, with some jumps for multi-byte characters, so an 8 byte character at `0` would jump to `8` for the next character.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -908,6 +911,74 @@ abstract class TextInputClient {
+  /// The position of this selection rect within the text.
+  final int position;
+
+  /// The rectangle representing the bounds of this selection rect.
```
Is it possible for the selection to be on multiple lines?  If so, is this like a bounding box?

fbcouch replied:
No, this is not itself a selection, it just represents a bounding box around a given character within a string of text, so the string "abc" will have selection rects for "a", "b", and "c".

LongCatIsLooong replied:
From the code in editable_text.dart, this in the current focused `RenderEditable`'s coordinate space?

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -908,6 +911,74 @@ abstract class TextInputClient {
+  /// Requests that the client show the editing toolbar, for example when the
+  /// platform changes the selection through a non-flutter method such as
+  /// scribble.
+  void showToolbar();
```
I worry about adding one more way to show the toolbar, since it's already pretty complicated figuring out what's causing it to show or not.  It does seem to be necessary in this case though...

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2431,13 +2439,60 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+  void _updateSelectionRects({bool force = false}) {
+    if (defaultTargetPlatform != TargetPlatform.iOS)
+      return;
+    // This is to avoid sending selection rects on non-iPad devices
```
Period at the end.

justinmc replied:
I guess we have no better way to determine if this is iPadOS...  At least we should make `1536.0` a constant called `_kIPadWidth` or something like that.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2431,13 +2439,60 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+    final Rect? firstRect = firstSelectionBoxes.isNotEmpty ? firstSelectionBoxes.first : null;
+    final ScrollDirection scrollDirection = _scrollController?.position.userScrollDirection ?? ScrollDirection.idle;
+    final Size size = renderEditable.size;
+    if (scrollDirection == ScrollDirection.idle && (force || text != _cachedText ||
+        _cachedFirstRect != firstRect || _cachedSize != size ||
+        _cachedPlaceholder != _placeholderLocation)) {
```
Would it make this any more readable if you split out some variable(s) out of this big conditional?

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -1502,6 +1502,239 @@ void main() {
+    expect(textSpan.text, 'Lorem ipsum dolor sit amet');
+  }, skip: kIsWeb);
+
+
```
Remove this extra indent.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -1502,6 +1502,239 @@ void main() {
+    textSpan = findRenderEditable(tester).text!;
+    expect(textSpan.children, null);
+    expect(textSpan.text, 'Lorem ipsum dolor sit amet');
+  }, skip: kIsWeb);
```
Should some of these tests be TargetPlatform.iOS only?

justinmc reviewed (commented):
Sorry for the slow response, I'm still really excited for Flutter to support this.  No major blockers in my comments, but we should also get @LongCatIsLooong to sign off.

Am I able to try this out using the simulator on a Mac, or do I need an iPad?

fbcouch commented:
@justinmc Thanks for your comments, I'll take a look through those individually – I have some memory management stuff to clean up on the engine side as well.

Yes, unfortunately there's no simulator support, so you will need a physical iPad with iPadOS 14 and a Pencil.

fbcouch commented:
@justinmc I think I made all the changes you requested / responded to questions. Let me know if I missed anything

blasten commented:
@fbcouch Awesome work! I was wondering if there's a design doc that provides an overview of this feature?  framework <-> engine interactions, etc...

fbcouch commented:
@blasten There isn't that I know of, but I can try to put one together if that would be helpful

blasten commented:
It would definitely help for future considerations. We have a template on https://flutter.dev/go/template

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2942,3 +3048,123 @@ class _Editable extends MultiChildRenderObjectWidget {
   }
 }
+
+class _ScribbleFocusable extends StatefulWidget {
```
Maybe in the long term we can make this public.  I think some users have built rich text editors that don't use EditableText, and they probably want Scribble support ideally.

CC @Renzo-Olivares

justinmc reviewed (commented):
Another round of mostly small cleanup comments.  Thanks for your persistence on this, I think it's almost ready to go.

fbcouch commented:
@blasten 
> It would definitely help for future considerations. We have a template on https://flutter.dev/go/template

All right, I have a first draft of that here: https://docs.google.com/document/d/1mjQbsSRQnHuAgMNdouaSgTS-Xv-w57fdKfOUUafWpRo

blasten commented:
Awesome! Thanks for sharing

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -911,6 +914,75 @@ abstract class TextInputClient {
+  /// For example, this method is called during a UIIndirectScribbleInteraction.
+  void onScribbleFocus(Offset offset);
+
+  /// Tests whether the [ScribbleClient] overlaps the given rectangle bounds.
```
It looks like it's exclusive in the default implementation that I provided using `Rect.overlaps`, but I'm not actually sure how PencilKit decides which input to choose other than that it's based on the bounds. I can add a note about that if it makes sense to do so, though.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2451,13 +2461,62 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     updateKeepAlive();
   }
 
+  String _cachedText = '';
+  Rect? _cachedFirstRect;
+  Size _cachedSize = Size.zero;
+  int _cachedPlaceholder = -1;
```
Yeah it seems like private members in this class get usually declared near where they're used. I can move it if it makes more sense to do so, but given how large this file is, it _might_ make more sense to be here? I don't have strong feelings about it either :smile:

justinmc replied:
Keeping it here sounds good 👍

fbcouch commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1311,10 +1402,45 @@ class TextInput {
+  @visibleForTesting
+  static Map<String, ScribbleClient> get scribbleClients => TextInput._instance._scribbleClients;
+
+  /// true if a scribble interaction is currently happening.
```
I added "Returns" that way I think we have the best of both

LongCatIsLooong commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -2794,6 +2794,14 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
+  ///
+  /// See [TextPainter.getBoxesForSelection] for more details.
+  List<Rect> getBoxesForSelection(TextSelection selection) =>
+    _textPainter.getBoxesForSelection(selection)
```
nit: this method needs to call `layoutText` before accessing the text metrics to ensure the text metrics are populated in `_textPainter`.

LongCatIsLooong replied:
nit: per the style guide https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-braces-for-long-functions-and-methods
> Use a block (with braces) when a body would wrap onto more than one line (as opposed to using ⇒; the cases where you can use ⇒ are discussed in the previous two guidelines).

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2451,13 +2461,62 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+        (int i) {
+          final int offset = _cachedText.characters.getRange(0, i).string.length;
+          return SelectionRect(
+            bounds: renderEditable.getBoxesForSelection(TextSelection(baseOffset: offset, extentOffset: offset + _cachedText.characters.characterAt(i).string.length)).first,
```
It looks like we can stop generating new `SelectionRect`s once it's completely below the current `RenderEditable`?

LongCatIsLooong replied:
Could there be performance issues if there's a fullscreen text field full of text? It's a bit worrying since iPads have large screens.

fbcouch replied:
Yes, some good points. I have been testing with a fairly long lorem text, and it seems to perform all right, but you might want to test as well.

It looks like the newline character selection box is currently at the end of the previous line. I'm not sure where it should be, though, but that seems like it makes sense.

I updated the generation code to stop once we're below the paintBounds of the renderEditable.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2739,6 +2828,23 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+      final List<_ScribblePlaceholder> placeholders = <_ScribblePlaceholder>[];
+      final int placeholderLocation = _value.text.length - _placeholderLocation;
+      if (_isMultiline) {
+        placeholders.add(_ScribblePlaceholder(child: Container(), size: Size.zero));
```
nit: Maybe use `SizedBox` to avoid the additional import?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1311,10 +1402,45 @@ class TextInput {
+        if (!(_scribbleClients[elementIdentifier]?.isInScribbleRect(rect) ?? false))
+          return false;
+        final Rect bounds = _scribbleClients[elementIdentifier]?.bounds ?? Rect.zero;
+        return !(bounds == Rect.zero || bounds.hasNaN || bounds.isInfinite);
```
:+1:

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2451,13 +2461,62 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
       SchedulerBinding.instance!
           .addPostFrameCallback((Duration _) => _updateSizeAndTransform());
+    } else if (_placeholderLocation != -1) {
+      SchedulerBinding.instance!
```
No, doesn't look like that's necessary – I took it out

justinmc commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -2725,6 +2725,16 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
+  List<Rect> getBoxesForSelection(TextSelection selection) {
+    _computeTextMetricsIfNeeded();
+    return _textPainter.getBoxesForSelection(selection)
+                       .map((TextBox textBox) => textBox.toRect().shift(_paintOffset))
+                       .toList();
```
Nit: Not sure if this is in the [styleguide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#indent-multi-line-argument-and-parameter-lists-by-2-characters) or not, so I could be wrong, but I thought we only indent by four characters here and don't align the dots.

fbcouch replied:
I can change it if you want, the closest thing I could find in the style guide was https://github.com/heartcombo/devise/blob/c82e4cf47b02002b2fd7ca31d441cf1043fc634c/lib/devise/models/recoverable.rb#L145, but it doesn't address this specific case

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2674,53 +2767,61 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
-                promptRectRange: _currentPromptRectRange,
-                promptRectColor: widget.autocorrectionTextRectColor,
-                clipBehavior: widget.clipBehavior,
+              child: _ScribbleFocusable(
```
Would it be worth it to exclude this widget from the tree on non-iPad devices?

fbcouch replied:
Maybe? It shouldn't add much in the way of overhead – just the map of active clients on  `TextInput` (`final Map<String, ScribbleClient> _scribbleClients = <String, ScribbleClient>{};`)

Let me know if you want me to exclude it, though, it shouldn't take much.

justinmc reviewed (commented):
There's an analyzer failure in a customer test, can you take a look and maybe just merge master into the PR to get it to go green again?

Otherwise just leaving a few quick comments as I catch up on the PR again.  Design doc looks great and helped me follow everything.

fbcouch commented:
@justinmc I merged in master, looks like customer_testing-linux is failing with a different problem now, looks to be something to do with version resolving...do you have any idea why that would break here but not on master (I _assume_ it is working on master)?

justinmc commented:
@fbcouch That failure seems to have nothing to do with the PR...  I tried rerunning it.  If it doesn't go green, could you try merging in master again?  Maybe you just got a bad commit or something.

fbcouch commented:
@justinmc I am at a loss for what that customer_testing-linux failure is about. I tried merging master a couple times, and still no luck

justinmc commented:
@fbcouch FYI there is a conflict now.

I've posted on hackers-infra on Discord to try to get some help with this failure.  I've seen similar failures, I don't think it's due to anything in this PR.  I have fixed similar failures by squashing all commits into one, but I don't want to make you do that unless we have to.

justinmc commented:
From Discord, it seems like the best solution here is to rebase with master.  That should get customer_testing-linux to pass.  Merge commits won't fix it.

fbcouch committed:
Implement framework side of iOS 14 Scribble

fbcouch commented:
@justinmc Thanks, that fixed the customer_testing-linux step! Unfortunately now some web tests are failing 😩 ...looks like some timeouts on dragging stuff, so not sure what that's about.

justinmc commented:
Everything is green now!  You've got some conflicts to fix though.

justinmc commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1129,6 +1129,16 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
+  /// Returns a list of rects that bound the given selection.
+  ///
+  /// See [TextPainter.getBoxesForSelection] for more details.
+  List<Rect> getBoxesForSelection(TextSelection selection) {
```
Would it be any cleaner if this method were added to [TextLayoutMetrics](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/text_layout_metrics.dart)?

The goal of that class is to simplify things by providing a read-only interface instead of the full RenderEditable.  You probably have to use RenderEditable anyway though, so maybe it's not worth it here.  Just wanted to bring it up.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -55,6 +56,9 @@ const Duration _kCursorBlinkWaitForStart = Duration(milliseconds: 150);
 // is shown in an obscured text field.
 const int _kObscureShowLatestCharCursorTicks = 3;
 
+// The minimum width of an iPad screen.
```
Could you add to the comment about where you got this value and what iPad models it's for?  I'm thinking it would be helpful in case a new iPad comes out that would require this constant to be changed.

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2523,20 +2527,84 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
     updateKeepAlive();
   }
 
+  String _cachedText = '';
```
Could storing this negatively affect performance if the field contains a large amount of text?

justinmc commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2657,6 +2726,30 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+  /// text.
+  ///
+  /// A value of -1 indicates there should be no placeholder, otherwise the
+  /// value should be between 0 and the length of the text, inclusive.
```
These triple slashes should be double I think.

justinmc commented on packages/flutter_test/lib/src/test_text_input.dart:
```diff
@@ -268,4 +270,98 @@ class TestTextInput {
   }
+
+  /// Simulates a scribble interaction starting
+  Future<void> startScribbleInteraction() async {
```
This is cool, I'm glad we can sort of fake these scribble interactions 👍

justinmc reviewed (commented):
Thanks for fixing the conflicts and failures, looks like everything is green.  I think this PR is close.

I want to make sure somebody can actually test how it works besides you. 
 @LongCatIsLooong Do you have an iPad and Apple Pencil that you could use to test this out with?  I have an iPad but no pencil...

fbcouch committed:
Update minimum iPad width for iPad Mini 6th Gen

fbcouch committed:
Fix up scribble methods on TestTextInput

skia-gold commented:
Gold has detected about 3 new digest(s) on patchset 43.
View them at https://flutter-gold.skia.org/cl/github/75472

fbcouch committed:
Add focus listener to CupertinoTextField init state

justinmc commented:
The [engine PR](https://github.com/flutter/engine/pull/24224) is merged 🎉 .  Let's wait for that to be rolled into the framework and then this can be merged.

fbcouch commented:
@justinmc Great! Thanks for all your time giving feedback on this! (And @LongCatIsLooong and @goderbauer as well)

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1102,6 +1105,76 @@ abstract class TextInputClient {
+/// This is used to report the current text selection rect and position data
+/// to the engine for Scribble support on iPadOS 14.
+@immutable
+class SelectionRect {
```
Could this be private?

fbcouch replied:
It's used in `EditableText` (https://github.com/flutter/flutter/pull/75472/files/7e28e92a313677a5aa17c8878c0b50eae49789d9#diff-f5f93c879cef9a102adbf148583de9f3d9a05f9678fdcad7bde106f62200bb34R2694 ) so I think it needs to be public. It would also be useful to be public if people are making custom text widgets that they want to be used with Scribble.

justinmc replied:
Ah I missed that, sounds good.

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -1271,6 +1351,18 @@ class TextInputConnection {
+    if (!listEquals(_cachedSelectionRects, selectionRects)) {
+      _cachedSelectionRects = selectionRects;
+      TextInput._instance._setSelectionRects(selectionRects.map((SelectionRect rect) {
+        return <num>[rect.bounds.left, rect.bounds.top, rect.bounds.width, rect.bounds.height, rect.position];
```
Could `num` be `double` to be more specific?

fbcouch replied:
We could probably make it work, but `rect.position` is an `int`, while the bounds properties are `double`, so I think `num` makes the most sense

justinmc reviewed (commented):
@fbcouch Quick question below about making SelectionRect private before I approve this.

The engine side has been rolled into the framework successfully now.

Also CC @LongCatIsLooong for second approval.

justinmc commented on packages/flutter_test/lib/src/test_text_input.dart:
```diff
@@ -271,4 +273,84 @@ class TestTextInput {
+    );
+  }
+
+  /// Simulates iOS removing a UITextPlaceholder after a long press with the pencil is released
```
Nit: Period at the end of all of these docs comments in this file.

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -1662,6 +1662,249 @@ void main() {
+    List<List<dynamic>> elements = await tester.testTextInput.scribbleRequestElementsInRect(const Rect.fromLTWH(0, 0, 1, 1));
+    expect(elements.first, containsAll(elementEntry));
+
+    // Touch is outside the bounds of the widget
```
Nit: Some missing periods in this file too.

justinmc reviewed (approved):
LGTM 👍

fbcouch committed:
Fix up comments

gaaclarke commented:
Friendly ping @justinmc, this was approved 20 days ago but hasn't been merged.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2649,20 +2653,84 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+    if (WidgetsBinding.instance!.window.physicalSize.shortestSide < _kIPadWidth)
+      return;
+    final String text = buildTextSpan().toPlainText(includeSemanticsLabels: false, includePlaceholders: false);
+    final List<Rect> firstSelectionBoxes = renderEditable.getBoxesForSelection(const TextSelection(baseOffset: 0, extentOffset: 1));
```
It would be nice if the underlying `TextPainter` can inform you when the text layout changes. Not sure what the API should look like though.

fbcouch replied:
I added a check for the text style as well – I started to remove this check, but it should at least have some positioning info that should be helpful. That's a good call though.

LongCatIsLooong reviewed (commented):
LGTM with some nits and questions. I'll trying this out in a simple benchmark and see if there's performance implications. We can probably make the feature opt-in if calculating the bounding boxes of all visible glyphs turns out to be expensive. 

Also this is likely a breaking change since it changes the `TextInputClient` interface, so you might need to follow the breaking change process. Alternatively I'm currently working on a pull request to make some of the features (e.g. floating cursor) optional, so you might be able to avoid taking the breaking change process if you wait for that to land first, but you may need to make some changes to the existing code.

fbcouch commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -3060,6 +3161,23 @@ class EditableTextState extends State<EditableText> with AutomaticKeepAliveClien
+        placeholders.add(const _ScribblePlaceholder(child: SizedBox(), size: Size(100.0, 0.0)));
+      }
+      return TextSpan(style: widget.style, children: <InlineSpan>[
+          TextSpan(text: _value.text.substring(0, placeholderLocation)),
```
Is it needed? I assumed (perhaps wrongly) that the children would inherit the parent's style when left empty

fbcouch committed:
Make new TextInputClient members; invalidate selection rect cache when text style changes

fbcouch commented:
@LongCatIsLooong Thanks for taking a look. Sorry it took a bit to come back to this, but I think I've addressed all of your comments.

Regarding the breaking change process, I went ahead and added a default (empty) implementation for the new members, so those should be opt-in now for clients that want to implement Scribble (given that's an iPadOS-specific feature, I'm guessing there are plenty of people _not_ interested in that). Is that a reasonable solution?

gaaclarke commented:
friendly ping @LongCatIsLooong

LongCatIsLooong commented:
Just did some quick testing on an iPad Air 2. The `_updateSelectionRects` call seems to be slightly slower than paragraph layout but is still somewhat reasonable when there's very few characters in the text field:

| initial text | `_updateSelectionRects` roughly takes (ms) |
|-|-|
| 10 characters ("helloworld") | 0.6 |
| 100 characters  (10 "helloworld"s) | 5 |
| 1000 characters (100 "helloworld"s) | 60 |

In some edge cases this can cause janks I feel, for instance when there's a text field with autofocus on, and route transition animation to the route that contains the text field can be janky, especially when the text field has a lot of text.

LongCatIsLooong commented:
Is there a way for the developer to opt out of the feature? For instance say the app's targeted users' language isn't one of the supported languages (German, French, Spanish, Italian, and Portuguese atm), or the developer has performance concerns?

fbcouch commented:
@LongCatIsLooong There isn't currently. Do you have a suggestion on the best way to go about that? Perhaps a flag on `EditableText` to enable or disable scribble?

LongCatIsLooong commented:
Yeah I think that makes sense.

fbcouch committed:
Add scribbleEnabled parameter to EditableText, TextField, and CupertinoTextField to allow opting out of the scribble features

fbcouch commented:
@LongCatIsLooong I added that `scribbleEnabled` flag to `EditableText`, `TextField`, and `CupertinoTextField`. The checks are all failing on some jwt error right now, but hopefully those will pass once whatever the underlying issue is fixed.

fbcouch committed:
Retry checks

LongCatIsLooong commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -1210,6 +1211,13 @@ class EditableText extends StatefulWidget {
   final ScrollPhysics? scrollPhysics;
 
+  /// {@template flutter.widgets.editableText.scribbleEnabled}
+  /// Whether iOS 14 Scribble features are enabled for this widget.
```
nit: maybe mention this is only supported on iPad?

LongCatIsLooong reviewed (approved):
LGTM! Sorry for the delay. 

Regarding the failure, I had it the other day and rebasing the commit to head/master fixed it.

fbcouch committed:
Fix up after rebase

fbcouch committed:
Add note that scribbleEnabled only available on iPads

fbcouch commented:
@LongCatIsLooong Thanks! I added a note to that effect, and rebasing onto master fixed the checks!

LongCatIsLooong commented:
Let's try merging this!

zhoushuangjian001 commented:
Flutter 3.7 version, ipad device, any page can be triggered, apple pencil's handwriting function
